tests/integration/pilot/revisioncmd/revision_view_test.go

```diff
@@ -0,0 +1,360 @@
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package revisioncmd
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"path/filepath"
+	"strings"
+	"testing"
+	"time"
+
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"istio.io/api/label"
+	"istio.io/istio/istioctl/cmd"
+	"istio.io/istio/pkg/test/env"
+	"istio.io/istio/pkg/test/framework"
+	"istio.io/istio/pkg/test/framework/components/echo"
+	"istio.io/istio/pkg/test/framework/components/echo/echoboot"
+	"istio.io/istio/pkg/test/framework/components/istioctl"
+	"istio.io/istio/pkg/test/framework/components/namespace"
+)
+
+var (
+	manifestPath = filepath.Join(env.IstioSrc, "manifests")
+
+	componentMap = map[string]string{
+		"istiod":          "Pilot",
+		"ingress-gateway": "IngressGateways",
+		"egress-gateway":  "EgressGateways",
+	}
+
+	expectedComponentsPerRevision = map[string]map[string]bool{
+		"stable": {
+			"base":                          true,
+			"istiod":                        true,
+			"ingress:istio-ingressgateway":  true,
+			"ingress:istio-eastwestgateway": true,
+			"egress:istio-egressgateway":    true,
+		},
+		"canary": {
+			"istiod":                        true,
+			"ingress:istio-ingressgateway":  true,
+			"ingress:istio-eastwestgateway": true,
+			"egress:istio-egressgateway":    true,
+		},
+	}
+)
+
+type revisionResource struct {
+	ns   namespace.Instance
+	echo echo.Instance
+}
+
+func TestRevisionCommand(t *testing.T) {
+	framework.
+		NewTest(t).
+		RequiresSingleCluster().
+		Features("installation.istioctl.revision_centric_view").
+		Run(func(ctx framework.TestContext) {
+			skipIfUnsupportedKubernetesVersion(ctx)
+			istioCtl := istioctl.NewOrFail(ctx, ctx, istioctl.Config{})
+
+			revisions := []string{"stable", "canary"}
+			revResourceMap := map[string]*revisionResource{}
+			builder := echoboot.NewBuilder(ctx)
+			for _, rev := range revisions {
+				effectiveRev := rev
+				if rev == "default" {
+					effectiveRev = ""
+				}
+				revResourceMap[rev] = &revisionResource{
+					ns: namespace.NewOrFail(t, ctx, namespace.Config{
+						Prefix:   fmt.Sprintf("istioctl-rev-%s", rev),
+						Inject:   true,
+						Revision: effectiveRev,
+					}),
+				}
+				builder.With(&revResourceMap[rev].echo,
+					echo.Config{Namespace: revResourceMap[rev].ns})
+			}
+			builder.BuildOrFail(ctx)
+
+			// Wait some time for things to settle down. This is very
+			// important for gateway pod related tests.
+			time.Sleep(60 * time.Second)
```

Comment:
Isn't this timeout too long?

Reply:
It is not a timeout. It is waiting for the cluster to reach steady state. Anyways, it could be reduced. I kept it to 1 minute to be safe and reduce chances of flakiness.

Reply:
I'll reduce it to 30 seconds.

Reply:
Done.